### PR TITLE
add log_tuples function

### PR DESCRIPTION
**Pull request by kk1050**
# ARTIQ Pull Request

## Description of Changes
add log_tuples function which can write the runid, start_time and path to a file in the data directory

### Related Issue
 Log runid/start_time/path tuples to a file in the data directory #810 

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :sparkles: New feature |

### All Pull Requests

- [x] Use correct spelling and grammar.
- [x] Close/update issues.
- [x] Check the copyright situation of your changes and sign off your patches (`git commit --signoff`, see [copyright](../CONTRIBUTING.rst#copyright-and-sign-off)).

### Code Changes

- [x] Test your changes or have someone test them. Mention what was tested and how.

### Git Logistics

- [x] Split your contribution into logically separate changes (`git rebase --interactive`). Merge/squash/fixup commits that just fix or amend previous commits. Remove unintended changes & cleanup. See [tutorial](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase).
- [x] Write short & meaningful commit messages. Review each commit for messages (`git show`). Format:
  ```
  topic: description. < 50 characters total.
  add log_tuples function

  Longer description. < 70 characters per line
  ```

### Licensing

See [copyright & licensing for more info](https://github.com/m-labs/artiq/blob/master/CONTRIBUTING.rst#copyright-and-sign-off).
ARTIQ files that do not contain a license header are copyrighted by M-Labs Limited and are licensed under LGPLv3+.

### CODE REVIEW

File: artiq/master/worker_impl.py [hunk 1]
```diff
@@ -342,7 +347,7 @@ def write_results():
                 try:
                     exp_inst.analyze()
                     put_completed()
-                finally:
+                finally:
```
Comment (sbourdeauducq) [inline artiq/master/worker_impl.py:350]: clean up

File: artiq/master/worker_impl.py [hunk 2]
```diff
@@ -286,6 +286,10 @@ def write_results():
             f["run_time"] = run_time
             f["expid"] = pyon.encode(expid)
 
+    def log_tuples():
+        with open("tuples.txt", "a") as f:
+            f.write("%s %s %s\n" %(rid, start_time, expid["file"]))
```
Comment (sbourdeauducq) [inline artiq/master/worker_impl.py:291]: use new style format ``"".format()"

**Comment by sbourdeauducq** [comment]
"tuples" is too generic of a name. https://en.wikipedia.org/wiki/Tuple

**Comment by sbourdeauducq** [comment]
What happens if multiple experiments are executed at the same time? This can happen if they are scheduled in different pipelines. Can't that corrupt the file?
Might be safer/simpler to write to the file from the scheduler where operations are done sequentially.

File: artiq/master/worker_impl.py [hunk 3]
```diff
@@ -286,6 +286,10 @@ def write_results():
             f["run_time"] = run_time
             f["expid"] = pyon.encode(expid)
 
+    def log_tuples():
+        with open("tuples.txt", "a") as f:
```
Comment (sbourdeauducq) [inline artiq/master/worker_impl.py:290]: The current directory of the experiment may be subject to change (see discussions in https://github.com/m-labs/artiq/issues/1747, https://github.com/m-labs/artiq/issues/1543).
Would be more future-proof to put it where the master writes other files such as ``last_rid.pyon``.

File: artiq/master/scheduler.py [hunk 4]
```diff
@@ -122,6 +122,13 @@ def __init__(self, ridc, worker_handlers, notifier, experiment_db):
         self.notifier = notifier
         self.experiment_db = experiment_db
 
+    def log_tuples(self, rid, expid):
```
Comment (sbourdeauducq) [inline artiq/master/scheduler.py:125]: "Tuples" is not a good name in functions either. What about ``log_submissions``?

File: artiq/frontend/artiq_master.py [hunk 5]
```diff
@@ -53,7 +53,10 @@ def get_argparser():
         "--experiment-subdir", default="",
         help=("path to the experiment folder from the repository root "
               "(default: '%(default)s')"))
-
+    group.add_argument(
+        "--log-filename", default="experiment_submissions.csv",
```
Comment (jordens) [inline artiq/frontend/artiq_master.py:57]: Please keep disabled by default

Comment (jordens) [inline artiq/frontend/artiq_master.py:57]: Please keep disabled by default

File: artiq/frontend/artiq_master.py [hunk 6]
```diff
@@ -54,9 +54,8 @@ def get_argparser():
         help=("path to the experiment folder from the repository root "
               "(default: '%(default)s')"))
     group.add_argument(
-        "--log-filename", default="experiment_submissions.csv", 
-        help=("set the filename of the experiment subimission "
-            "(default: '%(default)s')"))
+        "--log-filename", default=None, 
+        help=("(optional) set the filename to create the experiment subimission "))
```
Comment (sbourdeauducq) [inline artiq/frontend/artiq_master.py:58]: grammar, typo, space at the end

Comment (sbourdeauducq) [reply artiq/frontend/artiq_master.py:58]: And I don't think we want this in the "repository" options group.

File: artiq/master/scheduler.py [hunk 7]
```diff
@@ -113,14 +114,24 @@ async def build(self):
...
 
         self.ridc = ridc
         self.worker_handlers = worker_handlers
         self.notifier = notifier
         self.experiment_db = experiment_db
+        self.log_filename = log_filename
+
+    def log_submissions(self, rid, expid):
+        self.rid = rid
+        self.expid = expid
```
Comment (sbourdeauducq) [inline artiq/master/scheduler.py:129]: Why?

File: artiq/master/scheduler.py [hunk 8]
```diff
@@ -113,14 +114,24 @@ async def build(self):
...
+        start_time = time()
```
Comment (sbourdeauducq) [inline artiq/master/scheduler.py:130]: move into ``if``

File: artiq/master/scheduler.py [hunk 9]
```diff
@@ -135,6 +146,7 @@ def submit(self, expid, priority, due_date, flush, pipeline_name):
             wd, repo_msg = None, None
         run = Run(rid, pipeline_name, wd, expid, priority, due_date, flush,
                   self, repo_msg=repo_msg)
+        self.log_submissions(rid, expid)
```
Comment (sbourdeauducq) [inline artiq/master/scheduler.py:149]: ``log_submission``, singular

File: artiq/test/test_scheduler.py [hunk 10]
```diff
@@ -90,7 +90,7 @@ def setUp(self):
 
     def test_steps(self):
         loop = self.loop
-        scheduler = Scheduler(_RIDCounter(0), dict(), None)
+        scheduler = Scheduler(_RIDCounter(0), dict(), None, "experiment_submissions.csv")
```
Comment (sbourdeauducq) [inline artiq/test/test_scheduler.py:93]: If you're not reading the file in the test, then don't write it and set it to ``None``. You could just use ``None`` as default value in the ``Scheduler``'s parameter, then it's backward-compatible and you don't need to edit the test and potentially other code.

File: artiq/frontend/artiq_master.py [hunk 11]
```diff
@@ -53,7 +53,9 @@ def get_argparser():
         "--experiment-subdir", default="",
         help=("path to the experiment folder from the repository root "
               "(default: '%(default)s')"))
-
+    group.add_argument(
+        "--log-filename", default=None,
```
Comment (sbourdeauducq) [inline artiq/frontend/artiq_master.py:57]: There are already other "logging" options in the master and this is confusing. ``--log-submissions``?

**Pull request merged**